### PR TITLE
20.1.0 cleanup

### PR DESCRIPTION
**Pull request by dmkozh**

### CODE REVIEW

File: soroban-env-common/src/symbol.rs [hunk 1]
```diff
@@ -167,21 +166,27 @@ impl SymbolSmall {
 
 impl Symbol {
     #[doc(hidden)]
-    pub const unsafe fn from_small_body(body: u64) -> Self {
-        // This is a helper that should only be used in macros where it's
-        // necessary to generate a Symbol constant with no possibility of error:
-        // it just forces the bits that might be set wrong (54 and 55) to zero.
-        let body = body & SMALL_MASK;
-        Symbol(SymbolSmall::from_body(body).0)
```
Comment (leighmcculloch) [inline soroban-env-common/src/symbol.rs:175]: What's the motivation for removing the `from_small_body` function? Unclear to me if it's related to the change, or a drive-by change.

File: soroban-env-common/src/symbol.rs [hunk 2]
```diff
@@ -565,26 +567,23 @@ mod test_without_string {
         let input = "stellar";
         let sym = SymbolSmall::try_from_str(input).unwrap();
         let sym_str = SymbolStr::from(sym);
-        let s: &str = sym_str.as_ref();
-        assert_eq!(s, input);
+        assert_eq!(sym_str.to_string(), input);
```
Comment (leighmcculloch) [inline soroban-env-common/src/symbol.rs:570]: Why is this change necessary? Did we remove the as_ref() fn? I don't see it removed.

File: soroban-env-common/src/symbol.rs [hunk 3]
```diff
@@ -224,40 +228,38 @@ impl<const N: u32> TryFrom<&StringM<N>> for SymbolSmall {
...
         let mut accum: u64 = 0;
-        while n < b.len() {
-            let ch = b[n] as char;
-            if n >= MAX_SMALL_CHARS {
-                return Err(SymbolError::TooLong(b.len()));
-            }
-            n += 1;
-            accum <<= CODE_BITS;
-            let v = match SymbolSmall::encode_char(ch) {
+        while n < bytes.len() {
```
Comment (leighmcculloch) [inline soroban-env-common/src/symbol.rs:255]: There's a behavior change in here that seems unnecessary. The length check is moved to be earlier so there are failure cases that might fail with a length check that would have previously failed with some other error if there were invalid bytes prior to the maximum length being reached. Is this behavior check unnecessary? If no, I think we could be surgical with this change so it's purely non-functional / refactor and has no impact on observable behavior where it doesn't need to.

There also seem to be other small changes to this function that also seem unnecessary, but probably have no functional impact, but it'd be easier to review this change if it was as minimal/surgical as possible. If the changes are required, no worries, it's just not obvious to me that they are.

File: soroban-env-common/src/val.rs [hunk 4]
```diff
@@ -609,18 +609,15 @@ impl Val {
...
+            // as their contents can't be `None`.
             ScVal::Vec(None) => return false,
             ScVal::Map(None) => return false,
-            ScVal::Vec(Some(v)) => return v.0.iter().all(|x| Val::can_represent_scval(x)),
-            ScVal::Map(Some(m)) => {
-                return m
-                    .0
-                    .iter()
-                    .all(|e| Val::can_represent_scval(&e.key) && Val::can_represent_scval(&e.val))
-            }
```
Comment (leighmcculloch) [inline soroban-env-common/src/val.rs:623]: What requires the changes to `can_represent_scval`? I'm admittedly not very familiar anymore with this area of the Val code, so my question here is likely due to that unfamiliarity rather than anything out of place.

File: soroban-env-host/src/host.rs [hunk 5]
```diff
@@ -804,15 +805,13 @@ impl EnvBase for Host {
...
-        for ch in s.chars() {
-            SymbolSmall::validate_char(ch)?;
-        }
-        let res = self.add_host_object(ScSymbol(
-            self.metered_slice_to_vec(s.as_bytes())?.try_into()?,
-        ));
+        let res = self.add_host_object(ScSymbol::try_from_bytes(
+            self,
+            self.metered_slice_to_vec(s)?.try_into()?,
+        )?);
```
Comment (leighmcculloch) [inline soroban-env-host/src/host.rs:814]: I noticed @graydon's patch he posted [here](https://stellarfoundation.slack.com/archives/C06ATA4RPRP/p1704397276350619) created a common way for all mem host object types to define themselves to be fallibly convertable from bytes. The change in this PR works too, and less abstractions ftw, but @graydon may have a reason for us introducing that abstraction and it may be valuable.

File: soroban-env-host/src/host/conversion.rs [hunk 6]
```diff
@@ -346,6 +342,23 @@ impl Host {
         })
     }
 
+    // Version of `to_host_val` for the internal cases where the value has to
+    // be valid by construction (e.g. read from ledger).
+    pub(crate) fn to_valid_host_val(&self, v: &ScVal) -> Result<Val, HostError> {
```
Comment (leighmcculloch) [inline soroban-env-host/src/host/conversion.rs:347]: I would have thought it'd just always be safer to do the check. Why do we need the optimization?

File: soroban-env-common/src/symbol.rs [hunk 7]
```diff
@@ -274,7 +276,7 @@ impl SymbolSmall {
     // This should not be generally available as it can easily panic.
     #[cfg(feature = "testutils")]
     pub const fn from_str(s: &str) -> SymbolSmall {
-        match Self::try_from_str(s) {
+        match Self::try_from_bytes(s.as_bytes()) {
```
Comment (leighmcculloch) [inline soroban-env-common/src/symbol.rs:279]: Change seems unnecessary, try_from_str still exists.

File: soroban-env-common/src/val.rs [hunk 8]
```diff
@@ -820,8 +817,12 @@ impl Debug for Val {
...
                 let ss: SymbolStr =
                     unsafe { <SymbolSmall as ValConvert>::unchecked_from_val(*self) }.into();
-                let s: &str = ss.as_ref();
-                write!(f, "Symbol({s})")
+                // Even though this may be called for an arbitrary, not necessarily well-formed
+                // `Val`, this is still safe thanks to `SymbolSmall` iteration implementation that
+                // only returns valid symbol characters or `\0` even for invalid bit
+                // representations.
+                let converted_str = unsafe { str::from_utf8_unchecked(ss.as_ref()) };
+                write!(f, "Symbol({})", converted_str)
```
Comment (leighmcculloch) [inline soroban-env-common/src/val.rs:825]: Is the introduction of the unsafe unchecked required here? Why can't we keep using the `let s: &str = ss.as_ref()`? I must be missing something because this is the second spot that looks like an `AsRef<&str>` might have been removed, but I can't see it being removed anywhere in this PR.

**Review by leighmcculloch — COMMENTED** [review]
Thanks for pursuing this. I think the internal cleanup of char -> u8 makes it a bit easier for me to understand the intent and so I think it's worthwhile. Some questions and comments inline.

File: soroban-env-common/src/env.rs [hunk 9]
```diff
@@ -168,7 +168,7 @@ pub trait EnvBase: Sized + Clone {
     fn string_new_from_slice(&self, slice: &[u8]) -> Result<StringObject, Self::Error>;
 
     /// Form a new `Symbol` host object from a slice of client memory.
-    fn symbol_new_from_slice(&self, slice: &str) -> Result<SymbolObject, Self::Error>;
+    fn symbol_new_from_slice(&self, slice: &[u8]) -> Result<SymbolObject, Self::Error>;
```
Comment (leighmcculloch) [inline soroban-env-common/src/env.rs:171]: If this change requires rewriting any macros upstream, I think we should add a u8 variant alongside the str variant rather than replacing. We're not removing str from the other fns like `map_new_from_slices`, so removing the str variant of symbol_new_ doesn't seem to achieve much from a holistic pov of this interface.

File: soroban-env-common/src/symbol.rs (hunk 2 above)
Comment (dmkozh) [reply soroban-env-common/src/symbol.rs:570]: I wanted to remove `as_ref`, but then we decided against that. I would probably still remove it though in order to reduce amount of unsafe code exported as it is not used in SDK and there is no way for the SDK user to get to `SymbolStr` (today it's probably fine, in 2 years someone introduces `SymbolStr` construction for whatever reason and suddenly `as_ref` becomes a footgun). In any case, even if we leave it, I would prefer to use the safer version in the env codebase.

File: soroban-env-common/src/symbol.rs (hunk 3 above)
Comment (dmkozh) [reply soroban-env-common/src/symbol.rs:255]: This doesn't change the observable behavior in host because it doesn't distinguish between symbol errors. I could revert this, but given that here we're changing the error enum itself (which is observable by a theoretical env-common user, which I'm not sure if exists), I thought we might as well make this function a bit more clear and have consistent symbol validation logic between short and long symbols (so that we always check len first, then characters).

File: soroban-env-common/src/val.rs (hunk 4 above)
Comment (dmkozh) [reply soroban-env-common/src/val.rs:623]: This function was fundamentally mis-used as it recursively walks `ScVal`, but is also being called for every `ScVal` instance, so we ended up with N^2 complexity of `ScVal` conversions (where N is the number of `ScVal`s in XDR). This function also doesn't use the depth checker, so is potentially unsafe (we didn't catch any stack overflows with tests, but it's still a potential issue).
The main logical change with this is that `to_host_val` should distinguish between the external and internal conversions and return internal errors only when we deal with the input that is guaranteed to be convertible.

File: soroban-env-host/src/host/conversion.rs (hunk 6 above)
Comment (dmkozh) [reply soroban-env-host/src/host/conversion.rs:347]: Which check? This is not an optimization, just an 'assertion'. It's not absolutely necessary, but seems pretty useful. Alternative is to just do all the conversions via `to_host_val`, which has a risk of not noticing a real internal error.

File: soroban-env-common/src/val.rs (hunk 8 above)
Comment (dmkozh) [reply soroban-env-common/src/val.rs:825]: Again, I still think we can remove it, just need to agree upon this. For now, I wanted to obviate the unsafe conversion usage (this is the only code path that actually requires is now). If we decide that it's important to keep `as_ref`, I'll revert this.

File: soroban-env-common/src/env.rs (hunk 9 above)
Comment (dmkozh) [reply soroban-env-common/src/env.rs:171]: It doesn't require any macro changes, just one respective change in the SDK implementation of `EnvBase`. SDK itself just uses `Symbol::try_from_val` in its wrapper.
I'm not strictly opposed to keeping 2 versions, but since one of them is in fact redundant, I would like to get rid of unnecessary maintenance burden and keep just one version. Again, if we're worried about hypothetical alternative SDK that uses this, I would rather only keep the old version (again, maintaining 2 versions seems unnecessary and will cause more confusion in the future).

File: soroban-env-common/src/symbol.rs (hunk 2 above)
Comment (leighmcculloch) [reply soroban-env-common/src/symbol.rs]: `SymbStr`'s `ToString` impl calls the `AsRef<str>` impl, so I don't think it's any safer as far as I can tell. For example:
https://github.com/stellar/rs-soroban-env/blob/75a838b4780890aa679bea3b6a47a0a029021d90/soroban-env-common/src/symbol.rs#L388-L391

File: soroban-env-host/src/host/conversion.rs [hunk 10]
```diff
@@ -500,7 +501,13 @@ impl Host {
             }
             ScVal::Bytes(b) => Ok(self.add_host_object(b.metered_clone(self)?)?.into()),
             ScVal::String(s) => Ok(self.add_host_object(s.metered_clone(self)?)?.into()),
-            ScVal::Symbol(s) => Ok(self.add_host_object(s.metered_clone(self)?)?.into()),
```
Comment (sisuresh) [inline soroban-env-host/src/host/conversion.rs:493]: I'm curious - What path was this code previously taking? It's a little hard to follow.

Comment (dmkozh) [reply soroban-env-host/src/host/conversion.rs:493]: There is actually no additional code path here - `add_host_object` is infallible, so it used to just create a symbol object unconditionally.

File: soroban-env-common/src/val.rs (hunk 8 above)
Comment (dmkozh) [reply soroban-env-common/src/val.rs]: I reverted the changes related to `as_ref` for now.

File: soroban-env-common/src/symbol.rs (hunk 2 above)
Comment (dmkozh) [reply soroban-env-common/src/symbol.rs]: We could use the `&[u8]` version, but I've just reverted the changes for the sake of simplicity.

File: soroban-env-host/src/host.rs [hunk 11]
```diff
@@ -2254,7 +2265,12 @@ impl VmCallerEnv for Host {
         let scv = self.visit_obj(b, |hv: &ScBytes| {
             self.metered_from_xdr::<ScVal>(hv.as_slice())
         })?;
-        if Val::can_represent_scval(&scv) {
+        // Metering bug: the representation check is not metered,
```
Comment (graydon) [inline soroban-env-host/src/host.rs:2268]: Could you file a bug on this for future consolidation?

File: soroban-env-host/src/host/conversion.rs [hunk 12]
```diff
@@ -490,7 +501,13 @@ impl Host {
...
+            ScVal::Symbol(s) => Ok(self
+                .add_host_object(ScSymbol::try_from_bytes(
```
Comment (graydon) [inline soroban-env-host/src/host/conversion.rs:505]: Could you add a comment here explaining why this is necessary so future refactors / maintainers don't remove it?

File: soroban-env-host/src/host/conversion.rs [hunk 13]
```diff
@@ -158,7 +159,6 @@ impl Host {
         durability: ContractDataDurability,
     ) -> Result<Rc<LedgerKey>, HostError> {
         let key_scval = self.from_host_val(k)?;
-        self.check_val_representable_scval(&key_scval)?;
```
Comment (graydon) [inline soroban-env-host/src/host/conversion.rs:161]: Why are these being removed?

File: soroban-env-host/src/host/conversion.rs [hunk 14]
```diff
@@ -332,9 +332,6 @@ impl Host {
 
     pub(crate) fn to_host_val(&self, v: &ScVal) -> Result<Val, HostError> {
         let _span = tracy_span!("ScVal to Val");
-        // This is an internal consistency check: this is an internal method and
-        // any caller should have previously rejected non-representable ScVals.
-        self.check_val_representable_scval(&v)?;
```
Comment (graydon) [inline soroban-env-host/src/host/conversion.rs:337]: Why is this being removed?

File: soroban-env-macros/src/synth_linear_memory_tests.rs [hunk 15]
```diff
@@ -245,7 +246,12 @@ pub fn generate_wasm_module_with_preloaded_linear_memory(
                     // prefill the last 256 bytes with some values
                     // push in the following order: offset(d), val, length(n)
                     f3.i32_const(#DATA_SECTION_0_START as i32);
-                    f3.i32_const(7);
+                    if #hf_name_str == "symbol_new_from_linear_memory" {
```
Comment (graydon) [inline soroban-env-macros/src/synth_linear_memory_tests.rs:249]: I don't understand this change; can you put a comment to explain it?

File: soroban-env-host/src/host/conversion.rs (hunk 12 above)
Comment (dmkozh) [reply soroban-env-host/src/host/conversion.rs:507]: Done.

**Review by graydon — CHANGES_REQUESTED** [review]
Small number of nits / comment requests, basically looks ok.

File: soroban-env-host/src/host.rs (hunk 11 above)
Comment (dmkozh) [reply soroban-env-host/src/host.rs:2268]: Opened https://github.com/stellar/rs-soroban-env/issues/1318

File: soroban-env-host/src/host/conversion.rs (hunk 13 above)
Comment (dmkozh) [reply soroban-env-host/src/host/conversion.rs:161]: `from_host_val` performs exactly the same check already.

File: soroban-env-host/src/host/conversion.rs (hunk 14 above)
Comment (dmkozh) [reply soroban-env-host/src/host/conversion.rs:337]: I've decided to use this function for general conversions of potentially malformed data. The 'always valid' case is handled by `to_valid_host_val` function.

File: soroban-env-macros/src/synth_linear_memory_tests.rs (hunk 15 above)
Comment (dmkozh) [reply soroban-env-macros/src/synth_linear_memory_tests.rs:250]: Added comment

**Review by graydon — APPROVED** [review]
